### DISCUSSION TITLE
Just getting three metrics

### DISCUSSION THREAD

**Discussion by misterfu** [martin-helmich/prometheus-nginxlog-exporter#177] [Q&A] [answered]
Hey - thanks for this piece of software first ;-)

After setting up the exporter and connecting it successfully to Prometheus, I only see the following metrics in Prometheus (and also via accessing localhost:4040/metrics):

- [my_namespace]_http_response_count_total
- [my_namespace]_http_response_size_bytes
- [my_namespace]_parse_errors_total

Am I missing something in the configuration to get all the other nice metrics, e.g. http_response_time_seconds(*)? I'm using the (minimal) config from https://www.martin-helmich.de/en/blog/monitoring-nginx.html (of course with "source_files" pointing to the right NGINX access log, otherwise I would see no metrics at all, right?). 

The configured log format also perfectly matches my NGINX config:
`format = "$remote_addr - $remote_user [$time_local] \"$request\" $status $body_bytes_sent \"$http_referer\" \"$http_user_agent\" \"$http_x_forwarded_for\""`

**Comment by misterfu** [comment, marked as answer]
... just one more thing: I'm pretty sure the exporter cannot extract Response Times anywhere from the log, but I can't find a clue how to bring this into the NGINX log from the example config?

**Reply by stempler** [reply]
There is an example log configuration including request time and upstream response time at the end of [this section in the README](https://github.com/martin-helmich/prometheus-nginxlog-exporter#frequently-asked-questions).

**Reply by misterfu** [reply]
> There is an example log configuration including request time and upstream response time at the end of [this section in the README](https://github.com/martin-helmich/prometheus-nginxlog-exporter#frequently-asked-questions).

Awesome - thank you so much, @stempler. As so often, the answer is "RTFM"... I was just expecting that these NGINX variables were already covered by the HOWTO I quoted above.

I have now defined my own custom NGINX log format and it works.

**Comment by anhdz-2k1** [comment]
Hi I am installing and facing the same situation as you. If you have fixed this problem can you help me? My log format is:'

log_format upstream_time '$remote_addr - $remote_user [$time_local] '
                         '"$request" $status $body_bytes_sent '
                         '"$http_referer" "$http_user_agent"'
                         'rt=$request_time uct="$upstream_connect_time" uht="$upstream_header_time" urt="$upstream_response_time"'

**Reply by misterfu** [reply]
Hi @anhdz-2k1 -

This looks exactly like the nginx log format I have configured. Did you check your logfile if everything is fine there?

See a random line from my log file:

`xxx.xxx.xxx.xxx - - [07/Jan/2022:18:49:39 +0100] "GET /xxx.html HTTP/1.1" 200 21742 "https://www.xxx.de/" "Mozilla/5.0 (Linux; Android 11; SAMSUNG SM-A505FN) AppleWebKit/537.36 (KHTML, like Gecko) SamsungBrowser/16.0 Chrome/92.0.4515.166 Mobile Safari/537.36" rt="0.255" uct="0.000" uht="0.254" urt="0.255"`

**Reply by anhdz-2k1** [reply]
My log is in the same format as yours. I think I'm wrong because of my format in the .hcl file:

   format = "$remote_addr - $remote_user [$time_local] \"$request\" $status $body_bytes_sent \"$http_referer\" \"$http_user_agent\" 
    \"$http_x_forwarded_for\""

Does your format look like this?

**Reply by anhdz-2k1** [reply]
My config .hcl file looks like this:


listen {
  port = 4040
  address = "0.0.0.0"
}

namespace "toyotahungyen" {
  format = "$remote_addr - $remote_user [$time_local] \"$request\" $status $body_bytes_sent \"$http_referer\" \"$http_user_agent\" \"$http_x_forwarded_for\""
  source = {
    files = [
      "/var/log/nginx/toyota-hungyen.bizfly.site_access.log"
    ]
  }

}

namespace "toyotahaiduong" {
  source = {
    files = [
      "/var/log/nginx/toyota-haiduong.bizfly.site_access.log"
    ]
  }
  format = "$remote_addr - $remote_user [$time_local] \"$request\" $status $body_bytes_sent \"$http_referer\" \"$http_user_agent\" \"$http_x_forwarded_for\""
}

**Reply by misterfu** [reply]
Hi @anhdz-2k1 - sorry for the delay in coming back, this is how I have configured the format in my .hcl file:

`format = "$remote_addr - $remote_user [$time_local] \"$request\" $status $body_bytes_sent \"$http_referer\" \"$http_user_agent\" rt=\"$request_time\" uct=\"$upstream_connect_time\" uht=\"$upstream_header_time\" urt=\"$upstream_response_time\""`

So I guess the issue could be that your need to escape the quote signs.

**Reply by anhdz-2k1** [reply]
@misterfu 
I've tried your format and it still doesn't work. Can you give me a reference to hcl file format and complete log format configuration in nginx?

**Reply by misterfu** [reply]
Hi @anhdz-2k1,

My nginx log format is the same as yours, but here you go:

`    log_format  apm  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" '
                      'rt="$request_time" uct="$upstream_connect_time" uht="$upstream_header_time" urt="$upstream_response_time"';`

HCL configuration:

`listen {
     port = 4040
     address = "0.0.0.0"
}

namespace "xxx" {
     format = "$remote_addr - $remote_user [$time_local] \"$request\" $status $body_bytes_sent \"$http_referer\" \"$http_user_agent\" rt=\"$request_time\" uct=\"$upstream_connect_time\" uht=\"$upstream_header_time\" urt=\"$upstream_response_time\""
     source_files = ["/www/logs/xxx.log"]
     labels {
        app = "wordpress"
     }
}`

**Reply by anhdz-2k1** [reply]
@misterfu  Thank you very much, I have found the error and fixed it on our system. Thank you again for supporting me. Wishing you a lifetime of health and success in work and life <3
